### DISCUSSION TITLE
Tiny mallocs

### DISCUSSION THREAD

**Discussion by robertlipe** [jschwefel/esp-idf-ds3231#2] [Q&A]
Hi. I'm glancing through your submission. The code looks really good. Congrats!

I have a question about the style of programming as used at https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L218

You're calling malloc for one byte, initializing it to zero, passing it to a thing, and then freeing it.

Is there a reason that this can't just be const uint8_t data_wr = 0; ?

This same basic pattern appears several times in the code, such as

https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L246

That makes that a very expensive way to get a single byte holding a zero. :-)

One possible reason is if i2c_master_receive is doing DMA on that buffer and this is a device that can't DMA from the stack, but I don't see anything in https://docs.espressif.com/projects/esp-idf/en/v5.3/esp32/api-reference/peripherals/i2c.html?highlight=i2c_master_receive#_CPPv418i2c_master_receive23i2c_master_dev_handle_tP7uint8_t6size_ti that says that buffers CAN'T be on the stack. (I've not thoroughly reviewed I2C for this discussion)

You're clearly aware of the approach as it's used in https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L261

You have many malloc/callocs for RTC_DATA_LENGTH and smaller. These are so small that even on a FreeRTOS task, you can be pretty sure you can safely get 64 bytes of stack - heck, the call to calloc() which is going to call malloc and memset and such are likely to use 64 bytes of stack once alignment and saved registers and such are all considered. You have many tiny functions where just the weight of allocating, checking, handling the failure, and such dominate the function. Just making those stack variable would lighten them up if it's possible. These are all pretty likely to be nearly leaf nodes; I'd say that anything under a hundred bytes or so is fair game to just use stack unless it's one of the many cases you have that's returning what's clearly a buffer holding things.

As  note, byte_to_binary() isn't thread-safe. You're welcome to not care. Accepting a buffer pointer is the pretty traditional way to do that if you don't want to allocate it here yourself, but in C it's always a hazard to be sure that someone handed you a sensibly sized buffer.
Also, strcat() in there is pretty weird. It returns the pointer of the next destination, so you could build up the return with
char* p = b; before the loop and the inside the loop, something like
*p++ = (x & z) ? '1' : '0'. 

...but that might be unnecessary. If ESP-IDF is running a recent GNU Libc, %b is a supported printf specifier that honors widths and zero fills and all that goodness. Try it and see.

I'd probably find better names for gpio_task_example and friends. If someone ever profiles code and needs to finger these functions, it could be a goose chase finding that code. Keep it in the ds3231 "namespace" that you've consistently used elsewhere.

Please don't do this: https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L1003C1-L1004C1 Espressif's own "cute" message for a crash is confusing enough.

https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L1030 has a few typos for "battery".

Overall, there's not really much of a point here. I just wanted you to know your code was reviewed by an experienced programmer and I give it a thumbs up. Looks nice! Welcome.

**Comment by jschwefel** [comment]
> Hi. I'm glancing through your submission. The code looks really good. Congrats!
> 
> I have a question about the style of programming as used at
> 
> https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L218
> 
> You're calling malloc for one byte, initializing it to zero, passing it to a thing, and then freeing it.
> 
> Is there a reason that this can't just be const uint8_t data_wr = 0; ?
> 
> This same basic pattern appears several times in the code, such as
> 
> https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L246
> 
> That makes that a very expensive way to get a single byte holding a zero. :-)
> 
> One possible reason is if i2c_master_receive is doing DMA on that buffer and this is a device that can't DMA from the stack, but I don't see anything in https://docs.espressif.com/projects/esp-idf/en/v5.3/esp32/api-reference/peripherals/i2c.html?highlight=i2c_master_receive#_CPPv418i2c_master_receive23i2c_master_dev_handle_tP7uint8_t6size_ti that says that buffers CAN'T be on the stack. (I've not thoroughly reviewed I2C for this discussion)
> 
> You're clearly aware of the approach as it's used in
> 
> https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L261
> 
> You have many malloc/callocs for RTC_DATA_LENGTH and smaller. These are so small that even on a FreeRTOS task, you can be pretty sure you can safely get 64 bytes of stack - heck, the call to calloc() which is going to call malloc and memset and such are likely to use 64 bytes of stack once alignment and saved registers and such are all considered. You have many tiny functions where just the weight of allocating, checking, handling the failure, and such dominate the function. Just making those stack variable would lighten them up if it's possible. These are all pretty likely to be nearly leaf nodes; I'd say that anything under a hundred bytes or so is fair game to just use stack unless it's one of the many cases you have that's returning what's clearly a buffer holding things.
> 
> As note, byte_to_binary() isn't thread-safe. You're welcome to not care. Accepting a buffer pointer is the pretty traditional way to do that if you don't want to allocate it here yourself, but in C it's always a hazard to be sure that someone handed you a sensibly sized buffer. Also, strcat() in there is pretty weird. It returns the pointer of the next destination, so you could build up the return with char* p = b; before the loop and the inside the loop, something like *p++ = (x & z) ? '1' : '0'.
> 
> ...but that might be unnecessary. If ESP-IDF is running a recent GNU Libc, %b is a supported printf specifier that honors widths and zero fills and all that goodness. Try it and see.
> 
> I'd probably find better names for gpio_task_example and friends. If someone ever profiles code and needs to finger these functions, it could be a goose chase finding that code. Keep it in the ds3231 "namespace" that you've consistently used elsewhere.
> 
> Please don't do this: https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L1003C1-L1004C1 Espressif's own "cute" message for a crash is confusing enough.
> 
> https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L1030
> 
> has a few typos for "battery".
> Overall, there's not really much of a point here. I just wanted you to know your code was reviewed by an experienced programmer and I give it a thumbs up. Looks nice! Welcome.

Holy shit dude!
1.  Thanks! I appreciate the feedback.
2.  There is A LOT of feedback to respond to. 
3.  I see you made a PR. I will look at it.


So here goes to the reponses:

https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L218
I am doing that becuase `i2c_master_transmit_receive` calls for a a `uint8_t* write_buffer`. Did not really think too much about it, just did it that way. I will look in to replacing it with your suggestion. This is certainly an example of my newness showing.

https://github.com/jschwefel/esp-idf-ds3231/blob/56891467442492e8f98d978669c373c37d4738a7/esp-idf-ds3231.c#L246
This one is intentional. The DS3231 sends all of the registers during a read.  I think I am able to send it a register index and start reading from there, but it always sends all of the registers. I did this mainly because I was having a hell of a time reading and writing data data when I first started the project. This is a great candidate for more efficient code.

I will definately take a look at the various malloc/callocs. I am using them, for the same reason as above, the functions I am calling call for pointers. (Calloc was used primarily for debugging purposes. If I see a 0 that I was not expecting to be a 0, it raises a flag that something did not function properly.)

If I would have found the reference to `%b`, I never would have used the `byte_to_binary()`. But, I really needed it to help with debugging and all I was finding was "not supported" I will see if it works.

Yeh, I missed that one at the end. Using `gpio_task_example` was just plain lazy and I meant to change it. I deserve that one.

As for "DANGER", "Danger Will Robinson. DANGER!"..... I was particularly frustrated when I wrote that. I do not even remember WHY I was frustrated. All I can say is that the code was royally PI$$ING me off when I wrote it. Thought I got rid of all of those.
